callback.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2023 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dig
+
+// CallbackInfo contains information about a provided function or decorator
+// called by Dig, and is passed to a [Callback] registered with [WithCallback].
+type CallbackInfo struct {
+
+	// Name is the name of the function in the format:
+	// <package_name>.<function_name>
+	Name string
+
+	// Error contains the error returned by the [Callback]'s associated
+	// function, if there was one.
+	Error error
+}
+
+// Callback is a function that can be registered with a provided function
+// or decorator with [WithCallback] to cause it to be called after the
+// provided function or decorator is run.
+type Callback func(CallbackInfo)
+
+// WithCallback returns an option that can be used with [(*Container).Provide]
+// or [(*Container).Decorate] to have Dig call the passed in [Callback]
+// after the corresponding constructor or decorator finishes running.
+//
+// For example, the following prints a simple message after "myFunc" and
+// "myDecorator" finish running:
+//
+//	c := dig.New()
+//	myCallback := func(ci CallbackInfo) {
+//		var errorAdd string
+//		if ci.Error != nil {
+//			errorAdd = fmt.Sprintf("with error: %v", ci.Error)
+//		}
+//		fmt.Printf("%q finished%v", ci.Name, errorAdd)
+//	}
+//	c.Provide(myFunc, WithCallback(myCallback)),
+//	c.Decorate(myDecorator, WithCallback(myCallback)),
+//
+// See [CallbackInfo] for more info on the information passed to the [Callback].
+func WithCallback(callback Callback) withCallbackOption {
```

Comment:
I went ahead and made this change to resolve the lint error, but I'm also not sure if it's an anti-pattern. The one thing I can think of is that it's still not satisfying from an API documentation point of view. For example, I won't be able to go here: https://pkg.go.dev/go.uber.org/dig#ProvideOption and immediately see every possible `ProvideOption`, because this one won't be listed. But returning an unexported type doesn't solve that either.

I was trying to be conservative with API exposures, but perhaps two separate functions `func WithConstructorCallback(callback Callback) ProvideOption` and `func WithDecorateCallback(callback Callback) DecorateOption` really is the better option?

Reply:
I see. Yeah, having separate functions might be a better option. I'm interested in what @sywhang thinks about this.

Reply:
Let's not do this and prefer to declare separate Options. 
```
WithProviderCallback(CallbackInfo) ProvideOption
WithDecoratorCallback(CallbackInfo) DecorateOption
```

It's possible we may be able to clean this up using generics after we clean up the definition of `ProvideOption` and `DecorateOption` by making some changes to its unexported method, but as of now I don't think that's within the scope of this PR. 

In the future we could consider doing this using generics, and by then we still have the name `WithCallback` reserved to do so.